### PR TITLE
feat(skin): add a new css flex box based portlet layout

### PR DESCRIPTION
**Pull request by ChristianMurphy**
##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [x] documentation is changed or added
-   [x] view conforms with [WCAG 2.0 AA][]

##### Description of change

CSS Flex allows for fine grained control how how elements appear in one dimension (x or y), and gives course grain control over the appearance in the other dimension (y or x).
There is a absolute width flex layout where: the layout gives portlets a constant size, and requests that the flex give an even amount of space between portlets in a row (still allowing for a custom page gutter to be set).
The flex layout will pack as many portlets into a row as possible, portlets that do not fit are pushed to the next row.
If the screen size shrinks below the default size, each portlet will automatically switch to 100% screen width, to avoid overflow scroll or content clipping.

There is also a series of layouts that roughly match the existing bootstrap column layouts.
Where each portlet is given a percentage width, that is applied only on specific screensizes.
This is different from the existing bootstrap grid in that portlet height will be evened out across each row.

resolves [UP-4982](https://issues.jasig.org/browse/UP-4982)

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

**Comment by cousquer** [comment]
👍 I can not wait to see more ... 😃

**Comment by ChristianMurphy** [comment]
percentage-flex-flow

**Comment by drewwills** [comment]
## Update

I have enhancements to the REST API and to the options in the Customize Drawer essentially working locally.

@ChristianMurphy -- we've been talking about the use of Flex layouts with app launchers that (typically) have the `no-chrome` (hamburger) chrome style selected;  but we can't count on the idea that they will only be used this way.

Take a look at this screen capture...

need-padding

This image shows the (default) student user layout in `flexColumns=3`.

I think we need some padding/margin between the `<section>` elements in the Flex layouts -- portlets with regular chrome look "jammed together."

**Comment by ChristianMurphy** [comment]
Updated:

App Launchers look about the same:
screencapture-localhost-8080-uportal-f-admin-normal-render-up-1513961132889

Content Portlets now have spacing, and automatically stretch (additional vertical spacing on shorter portlets) to make the row sizes even.
screencapture-localhost-8080-uportal-f-u24l1s7-normal-render-up-1513961017262

screencapture-localhost-8080-uportal-f-welcome-normal-render-up-1513961004560

**Comment by drewwills** [comment]
@ChristianMurphy, @cousquer, et al. -- AFAIK, this update is complete and ready to merge.  I've done some local testing, but I'm planning to do a bit more between 2017/12/29 and 2018/01/02.  (I'm hoping to merge by Tuesday, Jan. 2nd at the latest.)

@ChristianMurphy -- I need a few more `.svg` icons.

File: uPortal-webapp/src/main/webapp/media/skins/respondr/common/images/layout_100.svg [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+<!-- Created with Inkscape (http://www.inkscape.org/) -->
+
+<svg
+   xmlns:dc="http://purl.org/dc/elements/1.1/"
+   xmlns:cc="http://creativecommons.org/ns#"
+   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
+   xmlns:svg="http://www.w3.org/2000/svg"
+   xmlns="http://www.w3.org/2000/svg"
+   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
+   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/webapp/media/skins/respondr/common/images/layout_100.svg]: :thinking: it looks like some inkscape graphics metadata leaked into the SVG.
This data is useful if the image needs to be edited again, but adds attributes that browsers ignore and there add a little unneeded weight when serving.
Should these be scrubbed?

File: uPortal-webapp/src/main/webapp/media/skins/common/javascript/uportal/up-layout-preferences.js [hunk 2]
```diff
@@ -60,14 +65,14 @@ var uportal = uportal || {};
...
         if (msg && type) {
-		var delay = (type == typeMsg.ERROR) ? 5000 : 2000;
-		if (messageDiv.length != 0) {
-			messageDiv.html('<p>' + msg + '</p>');
-			messageDiv.removeClass().addClass(type).show().delay(delay).fadeOut(400, callback);
-		} else return callback;
-	} else {
-		return callback;
-	}
+		        var delay = (type == typeMsg.ERROR) ? 5000 : 2000;
```
Comment (ChristianMurphy) [inline uPortal-webapp/src/main/webapp/media/skins/common/javascript/uportal/up-layout-preferences.js]: :art: thanks! :bow:

File: uPortal-webapp/src/main/webapp/media/skins/respondr/common/images/layout_100.svg (hunk 1 above)
Comment (ChristianMurphy) [reply uPortal-webapp/src/main/webapp/media/skins/respondr/common/images/layout_100.svg]: removed with [svgo](https://github.com/svg/svgo)

**Pull request merged**